src/algebra/direct_sum_graded.lean

```diff
@@ -0,0 +1,401 @@
+  gcomm_monoid (λ i, carriers i) :=
+gcomm_monoid.of_submonoids (λ i, (carriers i).to_add_submonoid) one_mem mul_mem
+
+/-- Build a `gcomm_monoid` instance for a collection of `submodules`s. -/
+def gcomm_monoid.of_submodules {R A : Type*}
+  [comm_semiring R] [comm_semiring A] [algebra R A] [add_comm_monoid ι]
+  (carriers : ι → submodule R A)
+  (one_mem : (1 : A) ∈ carriers 0)
+  (mul_mem : ∀ ⦃i j⦄ (gi : carriers i) (gj : carriers j), (gi * gj : A) ∈ carriers (i + j)) :
+  gcomm_monoid (λ i, carriers i) :=
+gcomm_monoid.of_submonoids (λ i, (carriers i).to_add_submonoid) one_mem mul_mem
+
+end shorthands
+
+/-! ### Instances for `⨁ i, A i` -/
+
+section one
+variables [has_zero ι] [ghas_one A] [∀ i, add_comm_monoid (A i)]
+
+instance : has_one (⨁ i, A i) :=
+{ one := direct_sum.of (λ i, A i) 0 ᵍ1}
+
+end one
+
+section mul
+variables [add_monoid ι] [∀ i, add_comm_monoid (A i)] [ghas_mul A]
+
+instance : has_mul (⨁ i, A i) :=
+{ mul := λ a b, begin
+    -- the elaborator struggles here, so use tactics to assemble the expression
+    refine direct_sum.to_add_monoid (λ j,
+      direct_sum.to_add_monoid (λ i,
+        (add_monoid_hom.comp_hom (direct_sum.of A $ i + j)).comp _
+      ) a
+    ) b,
+    exact (ᵍ*),
+  end }
+
+instance : mul_zero_class (⨁ i, A i) :=
+{ mul := (*),
+  zero := 0,
+  zero_mul := by { unfold has_mul.mul, simp [direct_sum.to_add_monoid, direct_sum.of] },
+  mul_zero := by { unfold has_mul.mul, simp [direct_sum.to_add_monoid, direct_sum.of] } }
+
+instance : distrib (⨁ i, A i) :=
+{ mul := (*),
+  add := (+),
+  left_distrib := by { unfold has_mul.mul, simp, },
+  right_distrib := by { unfold has_mul.mul, simp [direct_sum.to_add_monoid, direct_sum.of], } }
+
+end mul
+
+/-! The various semiring fields are proved separately because the proofs are slow. -/
+
+section semiring
+variables [∀ i, add_comm_monoid (A i)] [add_monoid ι] [gmonoid A]
+
+private lemma one_mul (x : ⨁ i, A i) : 1 * x = x :=
+begin
+  unfold has_one.one has_mul.mul,
+  simp only [add_monoid_hom.coe_mk, to_add_monoid_of, add_monoid_hom.comp_hom_apply_apply],
+  simp only [direct_sum.to_add_monoid, dfinsupp.lift_add_hom_apply, direct_sum.of],
+  convert add_monoid_hom.congr_fun (dfinsupp.sum_add_hom_single_add_hom) x,
+  ext1 i, ext1 xi,
+  exact dfinsupp.single_eq_of_sigma_eq (gmonoid.one_mul xi),
+end
+
+private lemma mul_one (x : ⨁ i, A i) : x * 1 = x :=
+begin
+  unfold has_one.one has_mul.mul,
+  simp only [add_monoid_hom.coe_mk, to_add_monoid_of, add_monoid_hom.comp_hom_apply_apply],
+  simp only [direct_sum.to_add_monoid, dfinsupp.lift_add_hom_apply, direct_sum.of],
+  rw add_monoid_hom.dfinsupp_sum_add_hom_apply x _,
+  convert add_monoid_hom.congr_fun (dfinsupp.sum_add_hom_single_add_hom ) x,
+  ext1 i, ext1 xi,
+  simp [dfinsupp.single_eq_of_sigma_eq (gmonoid.mul_one xi)],
+end
+
+private lemma mul_assoc (a b c : ⨁ i, A i) : a * b * c = a * (b * c) :=
+begin
```

Comment:
This proof looks really weird, did it come from `tidy`? Lots of `simp only`s can be combined (and above too), same with `rw`/`erw`s.

Reply:
Nope, it was written by hand...

Reply:
I think I found that sometimes `simp only [x, y], simp only [z]` would behave differently compared to ``simp only [x, y, z]`, since the former requires `z` to happen only after all x and y are done. I'll see if I can reduce this further though. Would you expect combining simp / erw / rw calls to speed up the proofs?

Reply:
I don't think combing calls would be expected to speed up a proof at all significantly. If that were the concern, possibly `simp_rw` with a correctly ordered list would be superior.

But I don't think speed is what Rob was noticing here; just that it's mildly confusing to have them separate (just because a reader can look at it and think "hmm, it's strange these were done as separate tactic invocations, there must be some good reason for that" and then be left unable to guess a reason).

Reply:
Something else that motivated it was to try and group together morally similar lemmas, so that the progression in tactic states looked reasonable - the depth of binders makes this proof really quite hard to keep track of.

Golfing attempts greatly appreciated, as usual!

Reply:
I've replaced almost all the `simp only`s with `rw`, and the proof is almost an order of magnitude faster.